### PR TITLE
Add array_max Spark Function

### PR DESCRIPTION
**Pull request by Ma-Jian1**
In presto, array_max returns NULL when input contains any NULL element.
In spark, NULL elements are skipped. If the array is empty or contains only NULL elements, NULL is returned.

Spark 3.2 ref: https://github.com/apache/spark/blob/v3.2.2/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala#L1852

### CODE REVIEW

File: velox/functions/sparksql/ArrayFunction.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  }
+};
+
+template <typename TExecCtx>
+struct ArrayMinFunction : public ArrayMinMaxFunction<TExecCtx, false> {};
+
+template <typename TExecCtx>
+struct ArrayMaxFunction : public ArrayMinMaxFunction<TExecCtx, true> {};
+
+} // namespace facebook::velox::functions
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/ArrayFunction.h:116]: Please add empty line to the end of file

File: velox/functions/sparksql/ArrayFunction.h [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cmath>
+#include <type_traits>
+#include "velox/functions/Udf.h"
+
+namespace facebook::velox::functions {
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/ArrayFunction.h:22]: Please use this namespace `facebook::velox::functions::sparksql` for sparksql function

File: velox/functions/sparksql/Register.cpp [hunk 3]
```diff
@@ -213,6 +220,18 @@ void registerFunctions(const std::string& prefix) {
...
+  registerArrayMinMaxFunctions<int16_t>(prefix);
+  registerArrayMinMaxFunctions<int32_t>(prefix);
+  registerArrayMinMaxFunctions<int64_t>(prefix);
+  registerArrayMinMaxFunctions<int128_t>(prefix);
+  registerArrayMinMaxFunctions<float>(prefix);
+  registerArrayMinMaxFunctions<double>(prefix);
+  registerArrayMinMaxFunctions<bool>(prefix);
+  registerArrayMinMaxFunctions<Varchar>(prefix);
+  registerArrayMinMaxFunctions<Timestamp>(prefix);
+  registerArrayMinMaxFunctions<Date>(prefix);
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/Register.cpp:234]: Do we need to register Date? Date is int32_t.

File: velox/functions/sparksql/tests/ArrayMinTest.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  testBool();
+}
+
+TEST_F(ArrayMinTest, numArrays) {
+  testNumNullable<float>();
+  testNumNullable<double>();
+}
+
+} // namespace
+} // namespace facebook::velox::functions::sparksql::test
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/tests/ArrayMinTest.cpp:199]: Same

File: velox/functions/sparksql/Register.cpp [hunk 5]
```diff
@@ -69,6 +70,12 @@ static void workAroundRegistrationMacro(const std::string& prefix) {
 
 namespace sparksql {
 
+template <typename T>
+inline void registerArrayMinMaxFunctions(const std::string& prefix) {
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/Register.cpp:74]: Why it is inline?

File: velox/functions/sparksql/ArrayFunction.h [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+      }
+    }
+  }
+
+  template <typename TReturn, typename TInput>
+  void assign(TReturn& out, const TInput& value) {
+    out = value;
+  }
+
+  void assign(out_type<Varchar>& out, const arg_type<Varchar>& value) {
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/ArrayFunction.h:61]: Does this `assign` function should also be inline as `FOLLY_ALWAYS_INLINE bool call`?

File: velox/functions/sparksql/ArrayFunction.h [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+      for (auto i = 1; i < array.size(); i++) {
+        update(currentValue, array[i].value());
+      }
+      assign(out, currentValue);
+      return true;
+    }
+
+    auto it = array.begin();
+    // NULL elements are skipped.
+    while (it != array.end() && !it->has_value()) {
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/ArrayFunction.h:88]: Can we optimize to only iterator the array once?
May like:
```
bool allIsNull = true;
while (it != array.end()) {
if (it->has_value()) {
alIsNull = false;
update(currentValue, it->value());
}
```

File: velox/functions/sparksql/ArrayFunction.h (hunk 2 above)
Comment (Ma-Jian1) [reply velox/functions/sparksql/ArrayFunction.h]: ok， I'll fix this in `array_min` pr, this pr depends on that

File: velox/functions/sparksql/ArrayFunction.h (hunk 6 above)
Comment (Ma-Jian1) [reply velox/functions/sparksql/ArrayFunction.h:61]: you can do it, but the compiler can't inline short functions like this.

File: velox/functions/sparksql/ArrayFunction.h (hunk 7 above)
Comment (Ma-Jian1) [reply velox/functions/sparksql/ArrayFunction.h:88]: then you can't get the `currentValue`
The current implementation ensures that the array is only iterated once.

File: velox/functions/sparksql/Register.cpp (hunk 5 above)
Comment (Ma-Jian1) [reply velox/functions/sparksql/Register.cpp:74]: this is copy from presto. There isn't much difference with or without `inline`.

File: velox/functions/sparksql/Register.cpp (hunk 3 above)
Comment (Ma-Jian1) [reply velox/functions/sparksql/Register.cpp:234]: This is copy from presto.
I think we should register for logical type.

File: velox/functions/sparksql/ArrayFunction.h (hunk 1 above)
Comment (Ma-Jian1) [reply velox/functions/sparksql/ArrayFunction.h]: fixed in array_min

File: velox/functions/sparksql/tests/ArrayMaxTest.cpp [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+TYPED_TEST(ArrayMaxFloatingPointTest, arrayMaxNullable) {
+  this->testWithNulls();
+}
+
+TYPED_TEST(ArrayMaxFloatingPointTest, arrayMax) {
+  this->testNoNulls();
+}
+
+} // namespace
+} // namespace facebook::velox::functions::sparksql::test
```
[the remark this answers is not in the document]
Comment (Ma-Jian1) [reply velox/functions/sparksql/tests/ArrayMaxTest.cpp]: done

File: velox/functions/sparksql/tests/ArrayMinTest.cpp (hunk 4 above)
Comment (Ma-Jian1) [reply velox/functions/sparksql/tests/ArrayMinTest.cpp]: done

**Comment by Ma-Jian1** [comment]
Close this due to merged to https://github.com/facebookincubator/velox/pull/6052

**Pull request closed without merging**